### ISSUE TITLE
Crash when debugging a test with Mock object

### ISSUE THREAD

**Issue by nerzhulart** [mono/mono#6348]
## Steps to Reproduce

1. Open solution in Rider 2017.3 (https://www.jetbrains.com/rider/download/)
[TestProject.tar.gz](https://github.com/mono/mono/files/1591709/TestProject.tar.gz)
2. Set breakpoint at SimpleTestClassTests.cs:17
3. Make step over

## Current Behavior

Mono.exe crashes

## Expected Behavior

Mono should not crash

### On which platforms did you notice this
[x] macOS
[ ] Linux 
[ ] Windows

**Version Used**:
Mac: Mono 5.4.0.199
### Stacktrace
```
/Library/Frameworks/Mono.framework/Versions/Current/bin/mono --debug --debugger-agent=transport=dt_socket,server=y,suspend=y,address=127.0.0.1:55555,setpgid=y "/Users/jetbrains/Library/Application Support/JetBrains/Toolbox/apps/Rider/ch-0/173.3801.283/Rider EAP.app/Contents/lib/ReSharperHost/JetBrains.ReSharper.TaskRunner.CLR45.x64.exe" 127.0.0.1 55917 89e5b524-bfda-4352-9875-7a82b5f3773a
mono(3529,0xb0577000) malloc: *** error for object 0x7902f0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Stacktrace:
Native stacktrace:
        0   mono                                0x00185a54 mono_handle_native_crash + 324
        1   mono                                0x001e5873 sigabrt_signal_handler + 147
        2   libsystem_platform.dylib            0xa1744e5b _sigtramp + 43
        3   ???                                 0xffffffff 0x0 + 4294967295
        4   libsystem_c.dylib                   0xa15caa34 abort + 141
        5   libsystem_malloc.dylib              0xa16c18e7 free + 529
        6   mono                                0x0036b27a mono_error_cleanup + 122
        7   mono                                0x001bedf1 type_commands_internal + 6049
        8   mono                                0x001b16a1 debugger_thread + 4593
        9   mono                                0x00372bf0 inner_start_thread + 192
        10  libsystem_pthread.dylib             0xa174e047 _pthread_body + 184
        11  libsystem_pthread.dylib             0xa174df8f _pthread_body + 0
        12  libsystem_pthread.dylib             0xa174d84a thread_start + 34
Debug info from gdb:
mono_gdb_render_native_backtraces not supported on this platform, unable to find gdb or lldb
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================
```

**Comment by lambdageek** [comment]
Confirmed on mono 5.4.1.7

This looks like a `mono_error_set_*` call without a `strdup` of one of the arguments.

**Comment by lambdageek** [comment]
Also crashes in VS for Mac, by the way.  (You have to go into the non-public members of `this` (a `SimpleTestClassTests` instance) to get the crash.  Same stack trace.)

**Comment by lambdageek** [comment]
No crash with mono 5.11.0.137 (`master/e5246d94ab9`). Looks like we fixed this somewhere along the way.  I'm going to check if there's anything we need to backport...

**Comment by nerzhulart** [comment]
Seems it's the same as https://youtrack.jetbrains.com/v2/issue/RIDER-8809

**Comment by lambdageek** [comment]
Thanks @nerzhulart, this indeed looks like the same bug. It's since it's already filed as https://bugzilla.xamarin.com/show_bug.cgi?id=57133, I'm going to close this issue and continue on bugzilla.

**Comment by lambdageek** [comment]
Tracked this down with git bisect: this issue was fixed by https://github.com/mono/mono/commit/522c8b69f5a433f645d15dc4ebb6a25b8266b414

So this looks like a duplicate of [bug 57488](https://bugzilla.xamarin.com/show_bug.cgi?id=57488)

That bug is fixed on mono master, 2017-12 and 2017-10 branches.
Also cherrypicking that commit to 2017-06 fixes the crash in mono 5.4.
